### PR TITLE
Include shadowJar artifact to release workflow

### PR DESCRIPTION
**Pull request by GumpacG**
### Description
Included shadowJar artifact to release workflow and bumped to 1.4.0.1
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/release-drafter.yml [hunk 1]
```diff
@@ -21,7 +21,17 @@ jobs:
...
         run: |
-          ./gradlew publishPublishMavenPublicationToLocalRepoRepository && tar -C build -cvf artifacts.tar.gz repository
+          ./gradlew publishPublishMavenPublicationToLocalRepoRepository
+      - name: Build shadowJar and include in artifacts
+        run: |
+          ./gradlew shadowJar
+          JDBC_VERSION=$(./gradlew :properties | grep version | cut -d ' ' -f 2)
+          SHADOW_JAR=$(echo build/libs/opensearch-sql-jdbc-shadow-"$JDBC_VERSION".jar)
+          ARTIFACTS=$(echo build/repository/org/opensearch/driver/opensearch-sql-jdbc/"$JDBC_VERSION")
+          mv $SHADOW_JAR $ARTIFACTS
```
Comment (Yury-Fridlyand) [inline .github/workflows/release-drafter.yml]: You can simplify that
```suggestion
          mv build/libs/opensearch-sql-jdbc-shadow-*.jar build/repository/org/opensearch/driver/opensearch-sql-jdbc/*/
```
Just confirm (manually test) that corresponding directory is created.
You won't have jars of multiple versions in CI.

File: .github/workflows/sql-jdbc-test-and-build-workflow.yml [hunk 2]
```diff
@@ -15,24 +15,24 @@ jobs:
           - 17
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v3
+      - uses: actions/checkout@v3
```
Comment (Yury-Fridlyand) [inline .github/workflows/sql-jdbc-test-and-build-workflow.yml]: `git` says that there are only whitespace changes. Do we need them?

Comment (GumpacG) [reply .github/workflows/sql-jdbc-test-and-build-workflow.yml]: This was unintended. Will revert.

File: .github/workflows/release-drafter.yml (hunk 1 above)
Comment (GumpacG) [reply .github/workflows/release-drafter.yml]: Worked! Changed in https://github.com/Bit-Quill/sql-jdbc/pull/8/commits/f28d5adfdd5692ec74d177bb727a967f1abc4463

**Review by Yury-Fridlyand — APPROVED** [review]
I'm worrying that shadow jar would be published on maven too...

**Review by MaxKsyunz — COMMENTED** [review]
Do we need to increment the version at this point? 

Changes to Jenkins are necessary before we can release 1.4.0.1 so it includes shadow jar.

**Comment by GumpacG** [comment]
@MaxKsyunz what would the process be like? Merge shadowJar workflow changes, then Change Jenkins, then bump version?

@Yury-Fridlyand do we not want shadowJar on Maven?

**Comment by Yury-Fridlyand** [comment]
No idea, it is not harmful though.
Right, we have to update `jenkins/release.jenkinsFile` too and add there a task to upload shadowjar to artifacts. See how it works on ODBC driver.

**Comment by GumpacG** [comment]
Based on the Jenkins file in https://github.com/opensearch-project/sql-jdbc/blob/6cc8e0a3c7933f13c58e32cbc6da3bdb9fb0d066/jenkins/release.jenkinsFile#L13C13-L13C59,
I think it signs everything that is in `build/repository` which is where the shadowJar is being moved to with the changes in the PR. Is there a way to test this without actually publishing to Maven?

**Comment by Yury-Fridlyand** [comment]
I think no.

**Comment by MaxKsyunz** [comment]
Having the shadow jar on Maven Central would be confusing -- if you are getting a dependency from Maven Central, you can just as easily get its dependencies and retain control of which versions are used. If there's a CVE in one of the dependencies, without shadow jar a client application can force the build to use another version of the dependency, with shadow jar we would need to release a new version of the shadow jar with the CVE fixed.

Regarding artifacts, the easiest to follow solution would be to save shadow jar as a separate artifact and have a separate stage in Jenkins job to process it.

`standardReleasePipelineWithGenericTrigger` can accept the name of the artifact to download. See [here](https://github.com/opensearch-project/opensearch-build-libraries/blob/de8ea9024a1a95a72c668562cbdd13065bf1ed46/vars/standardReleasePipelineWithGenericTrigger.groovy#L4).

File: jenkins/release.jenkinsFile [hunk 3]
```diff
@@ -14,4 +14,18 @@ standardReleasePipelineWithGenericTrigger(
...
         )
+        publishToArtifactsProdBucket(
+            assumedRoleName: 'sql-jdbc-upload-role',
+            source: "$WORKSPACE/artifacts.tar.gz",
+            destination: "https://artifacts.opensearch.org/opensearch-clients/jdbc/opensearch-sql-jdbc-${tag}.tar.gz",
+            signingPlatform: 'windows',
+            sigOverwrite: true
+        )
+        publishToArtifactsProdBucket(
+            assumedRoleName: 'sql-jdbc-shadowJar-upload-role',
```
Comment (acarbonetto) [inline jenkins/release.jenkinsFile:25]: they have a separate role for shadowJars?

**Review by acarbonetto — APPROVED** [review]

File: jenkins/release.jenkinsFile (hunk 3 above)
Comment (GumpacG) [reply jenkins/release.jenkinsFile:25]: I'm not actually too sure this works. I assumed that it was just a name for the jenkins job.
`@param Map[assumedRoleName] <required> - IAM role to be assumed for uploading artifacts`
Is from https://github.com/opensearch-project/opensearch-build-libraries/blob/de8ea9024a1a95a72c668562cbdd13065bf1ed46/vars/publishToArtifactsProdBucket.groovy#L12C1-L12C88 where that is being called.

File: jenkins/release.jenkinsFile [hunk 4]
```diff
@@ -14,4 +14,18 @@ standardReleasePipelineWithGenericTrigger(
             mavenArtifactsPath: "$WORKSPACE/repository/",
             autoPublish: true
         )
+        publishToArtifactsProdBucket(
```
Comment (Yury-Fridlyand) [inline jenkins/release.jenkinsFile:17]: why do you have 2 `publishToArtifactsProdBucket` for one jar?
The second one seems more correct

File: jenkins/release.jenkinsFile [hunk 5]
```diff
@@ -14,4 +14,18 @@ standardReleasePipelineWithGenericTrigger(
...
+            assumedRoleName: 'sql-jdbc-upload-role',
+            source: "$WORKSPACE/artifacts.tar.gz",
+            destination: "https://artifacts.opensearch.org/opensearch-clients/jdbc/opensearch-sql-jdbc-${tag}.tar.gz",
+            signingPlatform: 'windows',
+            sigOverwrite: true
+        )
+        publishToArtifactsProdBucket(
+            assumedRoleName: 'sql-jdbc-shadowJar-upload-role',
+            source: "$WORKSPACE/build/libs/opensearch-sql-jdbc-shadow-*.jar",
```
Comment (Yury-Fridlyand) [inline jenkins/release.jenkinsFile]:
```suggestion
            source: "$WORKSPACE/opensearch-sql-jdbc-shadow-${tag}.jar",
```

File: jenkins/release.jenkinsFile [hunk 6]
```diff
@@ -14,4 +14,18 @@ standardReleasePipelineWithGenericTrigger(
...
+            destination: "https://artifacts.opensearch.org/opensearch-clients/jdbc/opensearch-sql-jdbc-shadow-${tag}.jar",
+            signingPlatform: 'windows',
```
Comment (Yury-Fridlyand) [inline jenkins/release.jenkinsFile]:
```suggestion
            signingPlatform: 'linux',
```

File: jenkins/release.jenkinsFile (hunk 4 above)
Comment (GumpacG) [reply jenkins/release.jenkinsFile:17]: Not sure how it works if we specify this. As I understand artifacts.tar.gz is uploaded by default. I'm not sure if it ignores this file if `publishToArtifactsProdBucket` is called.

File: jenkins/release.jenkinsFile (hunk 5 above)
Comment (GumpacG) [reply jenkins/release.jenkinsFile]: `build/libs/opensearch-sql-jdbc-shadow-*.jar` would be the location of where the shadowJar is generated when it is built.

Comment (Yury-Fridlyand) [reply jenkins/release.jenkinsFile]: Yes, but jenkins gets it from release artifacts.

Comment (GumpacG) [reply jenkins/release.jenkinsFile]: Ohhh I see. Thanks

**Review by MaxKsyunz — APPROVED** [review]

**Pull request merged**